url75694980d5_fabioalmeid_tp01ftc_tgz-pJ8-agentes_jade_Centralizador_ExemploParserGramaticaCentralizadorJ8/src/Centralizador/PrettyPrinter.java

```diff
@@ -0,0 +1,568 @@
+       if (_i_ > 0) render(_R_PAREN);
+    }
+    else     if (foo instanceof Centralizador.Absyn.EData5)
+    {
+       Centralizador.Absyn.EData5 _edata5 = (Centralizador.Absyn.EData5) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       pp(_edata5.dados_1, 0);
+       pp(_edata5.operador_, 0);
+       pp(_edata5.dados_2, 0);
+       if (_i_ > 0) render(_R_PAREN);
+    }
+  }
+
+  private static void pp(Centralizador.Absyn.Operador foo, int _i_)
+  {
+    if (foo instanceof Centralizador.Absyn.EOp)
+    {
+       Centralizador.Absyn.EOp _eop = (Centralizador.Absyn.EOp) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       render("e");
+       if (_i_ > 0) render(_R_PAREN);
+    }
+  }
+
+  private static void pp(Centralizador.Absyn.Medicacao foo, int _i_)
+  {
+    if (foo instanceof Centralizador.Absyn.EMedic1)
+    {
+       Centralizador.Absyn.EMedic1 _emedic1 = (Centralizador.Absyn.EMedic1) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       pp(_emedic1.quantidade_, 0);
+       render(" ml de ");
+       pp(_emedic1.remedio_, 0);
+       if (_i_ > 0) render(_R_PAREN);
+    }
+    else     if (foo instanceof Centralizador.Absyn.EMedic2)
+    {
+       Centralizador.Absyn.EMedic2 _emedic2 = (Centralizador.Absyn.EMedic2) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       pp(_emedic2.remedio_, 0);
+       if (_i_ > 0) render(_R_PAREN);
+    }
+  }
+
+  private static void pp(Centralizador.Absyn.Quantidade foo, int _i_)
+  {
+    if (foo instanceof Centralizador.Absyn.EQtde1)
+    {
+       Centralizador.Absyn.EQtde1 _eqtde1 = (Centralizador.Absyn.EQtde1) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       pp(_eqtde1.integer_, 0);
+       if (_i_ > 0) render(_R_PAREN);
+    }
+  }
+
+  private static void pp(Centralizador.Absyn.Remedio foo, int _i_)
+  {
+    if (foo instanceof Centralizador.Absyn.ERemedy1)
+    {
+       Centralizador.Absyn.ERemedy1 _eremedy1 = (Centralizador.Absyn.ERemedy1) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       render("Dipirona");
+       if (_i_ > 0) render(_R_PAREN);
+    }
+    else     if (foo instanceof Centralizador.Absyn.ERemedy2)
+    {
+       Centralizador.Absyn.ERemedy2 _eremedy2 = (Centralizador.Absyn.ERemedy2) foo;
+       if (_i_ > 0) render(_L_PAREN);
+       render("Paracetamol");
+       if (_i_ > 0) render(_R_PAREN);
+    }
+  }
+
+
+  private static void sh(Centralizador.Absyn.Tarefa foo)
+  {
+    if (foo instanceof Centralizador.Absyn.ETask)
+    {
+       Centralizador.Absyn.ETask _etask = (Centralizador.Absyn.ETask) foo;
+       render("(");
+       render("ETask");
+       sh(_etask.acao_);
+       render(")");
+    }
+  }
+
+  private static void sh(Centralizador.Absyn.Acao foo)
+  {
+    if (foo instanceof Centralizador.Absyn.EAction1)
+    {
+       Centralizador.Absyn.EAction1 _eaction1 = (Centralizador.Absyn.EAction1) foo;
+       render("(");
+       render("EAction1");
+       sh(_eaction1.coletar_);
+       sh(_eaction1.dados_);
+       render(")");
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 391:398, 400:407, 489:496.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.